### PR TITLE
ast,optimizer: switch to use new parser, and convert ast.Node to old statements.

### PR DESCRIPTION
**Pull request by shenli**

### CODE REVIEW

File: ast/base.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ func (dn *dmlNode) dmlStatement() {}
 // Expression implementations should embed it in.
 type exprNode struct {
```
Comment (shenli) [inline ast/base.go]: This will replace expression.Value?

File: ast/cloner.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+// Leave implements Visitor Leave interface.
+func (c *Cloner) Leave(in Node) (out Node, ok bool) {
+	return in, true
+}
+
+// cloneStruct clone a node's struct value, if the struct has slice
+// the cloned value should make a new slice and copy old slice to new slice.
+func cloneStruct(in Node) (out Node) {
```
Comment (siddontang) [inline ast/cloner.go]: Maybe we can use some DeepCopy package to simplify it.
Btw, why not supporting Clone interface?

File: ast/ddl.go [hunk 3]
```diff
@@ -524,23 +550,25 @@ type AlterTableStmt struct {
...
-func (at *AlterTableStmt) Accept(v Visitor) (Node, bool) {
-	if skipChildren, ok := v.Enter(at); skipChildren {
-		return at, ok
+func (nod *AlterTableStmt) Accept(v Visitor) (Node, bool) {
+	newNod, skipChildren := v.Enter(nod)
+	if skipChildren {
+		return v.Leave(newNod)
 	}
-	node, ok := at.Table.Accept(v)
+	nod = newNod.(*AlterTableStmt)
```
Comment (shenli) [inline ast/ddl.go]: nod?

File: ast/base.go (hunk 1 above)
Comment (coocood) [inline ast/base.go]: ValueExpr will replace expression.Value

File: ast/cloner.go (hunk 2 above)
Comment (coocood) [inline ast/cloner.go]: I think do the job in visitor is better, so we can keep the same logic in different node in the one place.

File: ast/ddl.go [hunk 4]
```diff
@@ -70,11 +70,13 @@ type CreateDatabaseStmt struct {
...
-func (cd *CreateDatabaseStmt) Accept(v Visitor) (Node, bool) {
-	if skipChildren, ok := v.Enter(cd); skipChildren {
-		return cd, ok
+func (nod *CreateDatabaseStmt) Accept(v Visitor) (Node, bool) {
+	newNod, skipChildren := v.Enter(nod)
+	if skipChildren {
+		return v.Leave(newNod)
 	}
-	return v.Leave(cd)
+	nod = newNod.(*CreateDatabaseStmt)
```
Comment (siddontang) [inline ast/ddl.go]: no need to check type?
the same for following type conversations.

File: ast/ddl.go (hunk 3 above)
Comment (coocood) [inline ast/ddl.go]: Name  'node' is taken for child element, so I use 'nod' for parent element.

File: ast/ddl.go (hunk 4 above)
Comment (coocood) [inline ast/ddl.go]: The returned type should be the same for Enter and Leave function. I'll add comments on Visitor methods.

File: ast/expressions.go [hunk 5]
```diff
@@ -47,21 +51,58 @@ var (
...
+	case nil:
+		ve.Type = types.NewFieldType(mysql.TypeNull)
+	case bool, int64:
+		ve.Type = types.NewFieldType(mysql.TypeLonglong)
+	case uint64:
+		ve.Type = types.NewFieldType(mysql.TypeLonglong)
+		ve.Type.Flag |= mysql.UnsignedFlag
+	case string, UnquoteString:
+		ve.Type = types.NewFieldType(mysql.TypeVarchar)
+		ve.Type.Charset = mysql.DefaultCharset
```
Comment (shenli) [inline ast/expressions.go]: Use charset/collation specified in character_set_connection and collation_connection system variables.

File: optimizer/evaluator.go [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/parser/opcode"
+	"github.com/pingcap/tidb/plan"
+)
+
+// Evaluator is a ast visitor that evaluates an expression.
+type Evaluator struct {
```
Comment (shenli) [inline optimizer/evaluator.go]: Why this in the optimizer package?

File: optimizer/validator.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package optimizer
+
+import "github.com/pingcap/tidb/ast"
+
+// validator is an ast.Visitor that validates
+// ast parsed from parser.
+type validator struct {
```
Comment (shenli) [inline optimizer/validator.go]: More comments?
I can not understand what is this used for.

**Comment by coocood** [comment]
Ready to merge.
PTAL @siddontang @shenli

File: ast/cloner.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+import "fmt"
+
+// Cloner is an ast visitor that clones a node.
+type Cloner struct {
+}
+
+// Enter implements Visitor Enter interface.
+func (c *Cloner) Enter(node Node) (Node, bool) {
+	return cloneStruct(node), false
```
Comment (shenli) [inline ast/cloner.go]: Why not skipChildren? I think all the work is done in Enter().

Comment (coocood) [inline ast/cloner.go]: Tthe returned new node is a cloned node, but its children are not,  its children need to be visited and newly cloned children are assigned back.

File: optimizer/infobinder.go [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+//
+// In general, a reference can only refer to information that are available for it.
+// So children elements are visited in the order that previous elements make information
+// available for following elements.
+//
+// During visiting, information are collected and stored in binderContext.
+// When we enter a sub query, a new binderContext is pushed to the contextStack, so sub query
+// information can overwrite outer query information. When we look up for a column reference,
+// we look up from top to bottom in the contextStack.
+type InfoBinder struct {
```
Comment (shenli) [inline optimizer/infobinder.go]: This is used to solve subquery?

Comment (coocood) [inline optimizer/infobinder.go]: It can not be used yet.

File: ast/dml.go [hunk 10]
```diff
@@ -399,58 +568,68 @@ const (
...
 		if !ok {
-			return in, false
+			return nod, false
 		}
-		in.OnDuplicate[i] = node.(*Assignment)
+		nod.OnDuplicate[i] = node.(*Assignment)
 	}
-	if in.Select != nil {
-		node, ok := in.Select.Accept(v)
+	if nod.Select != nil {
```
Comment (shenli) [inline ast/dml.go]: visit this first

**Comment by shenli** [comment]
LGTM

File: ast/ast.go [hunk 11]
```diff
@@ -93,11 +102,12 @@ type ResultSetNode interface {
 // Visitor visits a Node.
 type Visitor interface {
 	// VisitEnter is called before children nodes is visited.
+	// The returned node should be the same type as the input node n.
```
Comment (siddontang) [inline ast/ast.go]: should be -> must be?

File: ast/ast.go [hunk 12]
```diff
@@ -26,6 +26,11 @@ type Node interface {
 	// Accept accepts Visitor to visit itself.
 	// The returned node should replace original node.
 	// ok returns false to stop visiting.
+	//
+	// Implementation of this method should first call visitor.Enter,
+	// assign the returned node to its method receiver, if skipChildren returns true,
```
Comment (qiuyesuifeng) [inline ast/ast.go]: I cannot see where is the variable of `skipChildren`, u mean `ok bool`?

Comment (coocood) [inline ast/ast.go]: skipChildren is returned by Visitor.Enter method

File: ast/ast.go [hunk 13]
```diff
@@ -26,6 +26,11 @@ type Node interface {
...
+	// children should be skipped. Otherwise, call its children in particular order that
+	// later elements depends on on former elements. Finally, return visitor.Leave.
```
Comment (qiuyesuifeng) [inline ast/ast.go]: `depends on on` should be `depends on`?

File: ast/cloner.go (hunk 2 above)
Comment (qiuyesuifeng) [inline ast/cloner.go]: How about combining the case conditions？
like
```
switch v := in.(type) {
      case *ValueExpr, *BetweenExpr, xxx:
            nv := *v
            out = &nv
}
```

Comment (coocood) [inline ast/cloner.go]: That doesn't compile.

File: ast/cloner.go [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+	case *PositionExpr:
+		nv := *v
+		out = &nv
+	case *PatternRegexpExpr:
+		nv := *v
+		out = &nv
+	case *RowExpr:
+		nv := *v
+		nv.Values = make([]ExprNode, len(v.Values))
+		copy(nv.Values, v.Values)
```
Comment (qiuyesuifeng) [inline ast/cloner.go]: Can copy function do a deep copy?

File: ast/cloner_test.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+	a1 := a.V
+	b1 := b.(*UnaryOperationExpr).V
+	c.Assert(a1, Not(Equals), b1)
+	a2 := a1.(*UnaryOperationExpr).V
+	b2 := b1.(*UnaryOperationExpr).V
+	c.Assert(a2, Not(Equals), b2)
+	a3 := a2.(*ValueExpr)
+	b3 := b2.(*ValueExpr)
+	c.Assert(a3, Not(Equals), b3)
+	c.Assert(a3.GetValue(), Equals, true)
```
Comment (qiuyesuifeng) [inline ast/cloner_test.go]: Please use IsTrue.

File: ast/ddl.go [hunk 16]
```diff
@@ -70,11 +70,13 @@ type CreateDatabaseStmt struct {
 }
 
 // Accept implements Node Accept interface.
-func (cd *CreateDatabaseStmt) Accept(v Visitor) (Node, bool) {
-	if skipChildren, ok := v.Enter(cd); skipChildren {
-		return cd, ok
+func (nod *CreateDatabaseStmt) Accept(v Visitor) (Node, bool) {
```
Comment (qiuyesuifeng) [inline ast/ddl.go]: I prefer node instead of nod, nod is a common word.

File: ast/cloner.go (hunk 14 above)
Comment (coocood) [inline ast/cloner.go]: No, just copy the pointer.

File: ast/ddl.go (hunk 16 above)
Comment (coocood) [inline ast/ddl.go]: 'node' is already used.

File: ast/cloner.go [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+	case *BinaryOperationExpr:
+		nv := *v
+		out = &nv
+	case *WhenClause:
+		nv := *v
+		out = &nv
+	case *CaseExpr:
+		nv := *v
+		nv.WhenClauses = make([]*WhenClause, len(v.WhenClauses))
+		copy(nv.WhenClauses, v.WhenClauses)
```
Comment (ngaut) [inline ast/cloner.go]: Do you wanna do deep copy ?

File: ast/cloner.go (hunk 2 above)
Comment (ngaut) [inline ast/cloner.go]: s/cloneStruct/copyStruct/g

File: ast/ddl.go [hunk 18]
```diff
@@ -474,44 +498,46 @@ type AlterTableSpec struct {
 	Constraint *Constraint
 	Options    []*TableOption
 	Column     *ColumnDef
-	ColumnName *ColumnName
+	DropColumn *ColumnName
```
Comment (qiuyesuifeng) [inline ast/ddl.go]: Why rename ColumnName to DropColumn?

File: ast/cloner.go (hunk 17 above)
Comment (coocood) [inline ast/cloner.go]: cloneStruct do shallow value copy and make new children slices

File: ast/dml.go [hunk 19]
```diff
@@ -399,58 +568,68 @@ const (
 type InsertStmt struct {
 	dmlNode
 
+	Replace     bool
```
Comment (qiuyesuifeng) [inline ast/dml.go]: I think insert and replace are different statement, so maybe it is not a good name.

File: ast/parser/parser.y [hunk 20]
```diff
@@ -431,15 +440,16 @@ import (
...
 	Operand			"operand"
 	OptFull			"Full or empty"
 	OptInteger		"Optional Integer keyword"
 	Order			"ORDER BY clause optional collation specification"
 	OrderBy			"ORDER BY clause"
-	OrderByItem		"ORDER BY list item"
+	ByItem			"BY item"
 	OrderByOptional		"Optional ORDER BY clause optional"
-	OrderByList 		"ORDER BY list"
+	ByList 			"BY list"
```
Comment (qiuyesuifeng) [inline ast/parser/parser.y]: Remove empty spaces.

File: optimizer/aggregator.go [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package optimizer
+
+// Aggregator is the interface to
+// compute aggregate function result.
+type Aggregator interface {
+	// Input add a input value to aggregator.
```
Comment (qiuyesuifeng) [inline optimizer/aggregator.go]: adds an input

File: optimizer/aggregator.go [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+	// The input values are accumulated in the aggregator.
+	Input(in ...interface{}) error
+	// Output use input values to compute the aggregated result.
```
Comment (qiuyesuifeng) [inline optimizer/aggregator.go]: uses

File: optimizer/compiler.go [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+// Compile compiles a ast.Node into an executable statement.
+func (com *Compiler) Compile(node ast.Node) (stmt.Statement, error) {
+	validator := &validator{}
+	if _, ok := node.Accept(validator); !ok {
+		return nil, errors.Trace(validator.err)
+	}
+
+	//	binder := &InfoBinder{}
```
Comment (qiuyesuifeng) [inline optimizer/compiler.go]: Can these lines be removed?

File: optimizer/convert_expr.go [hunk 24]
```diff
@@ -0,0 +1,425 @@
...
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/subquery"
+	"github.com/pingcap/tidb/model"
+	"strings"
+)
+
+func convertExpr(converter *expressionConverter, expr ast.ExprNode) (expression.Expression, error) {
+	expr.Accept(converter)
```
Comment (qiuyesuifeng) [inline optimizer/convert_expr.go]: Any need to process return values?

File: optimizer/compiler.go (hunk 23 above)
Comment (coocood) [inline optimizer/compiler.go]: It will be used later.

File: optimizer/evaluator.go [hunk 25]
```diff
@@ -0,0 +1,317 @@
...
+	ret.Accept(e)
+	return e.err == nil
+}
+
+func columnCount(e ast.ExprNode) (int, error) {
+	switch x := e.(type) {
+	case *ast.RowExpr:
+		n := len(x.Values)
+		if n <= 1 {
+			return 0, errors.Errorf("Operand should contain >= 2 columns for Row")
```
Comment (qiuyesuifeng) [inline optimizer/evaluator.go]: errors.New

File: optimizer/convert_expr.go (hunk 24 above)
Comment (coocood) [inline optimizer/convert_expr.go]: returned value is not needed.

File: optimizer/evaluator.go (hunk 25 above)
Comment (coocood) [inline optimizer/evaluator.go]: This will be replaced by terror later.

File: optimizer/infobinder.go [hunk 26]
```diff
@@ -0,0 +1,441 @@
...
+	// result fields collected in group by clause.
+	groupBy []*ast.ResultField
+
+	// The join node stack is used by on condition to find out
+	// available tables to reference. On condition can only
+	// refer to tables involved in current join.
+	joinNodeStack []*ast.Join
+
+	// When visiting TableRefs, tables in this context are not available
+	// because it is being collected.
```
Comment (qiuyesuifeng) [inline optimizer/infobinder.go]: How about combining `inXXX` into an const value?

File: optimizer/typecomputer.go [hunk 27]
```diff
@@ -13,20 +13,18 @@
...
 
-// Compile compiles a ast.Node into a executable statement.
-func Compile(node ast.Node) (stmt.Statement, error) {
-	switch v := node.(type) {
-	case *ast.SetStmt:
-		return compileSet(v)
-	}
-	return nil, nil
+// typeComputer is an ast Visitor that
+// Compute types for ast.ExprNode.
```
Comment (qiuyesuifeng) [inline optimizer/typecomputer.go]: computes

Comment (qiuyesuifeng) [inline optimizer/typecomputer.go]: And what is the meaning of `compute types`?

File: optimizer/validator.go (hunk 7 above)
Comment (qiuyesuifeng) [inline optimizer/validator.go]: +1

File: tidb.go [hunk 28]
```diff
@@ -132,15 +131,12 @@ func Compile(ctx context.Context, src string) ([]stmt.Statement, error) {
...
-		if node, ok := v.(ast.Node); ok {
-			stm, err := optimizer.Compile(node)
-			if err != nil {
-				return nil, errors.Trace(err)
-			}
-			stmts[i] = stm
-		} else {
-			stmts[i] = v.(stmt.Statement)
+		compiler := &optimizer.Compiler{}
+		stm, err := compiler.Compile(v)
```
Comment (qiuyesuifeng) [inline tidb.go]: Why not use stmt but stm?
For u have defined stmts in lint 132.

File: optimizer/typecomputer.go (hunk 27 above)
Comment (coocood) [inline optimizer/typecomputer.go]: The type of an expression is computed by its operands and operator.
the type of ( "1" + 1) is double.

File: tidb.go (hunk 28 above)
Comment (coocood) [inline tidb.go]: stmt is the package name.

File: optimizer/typecomputer.go (hunk 27 above)
Comment (qiuyesuifeng) [inline optimizer/typecomputer.go]: So u mean operation type?
computer is very strange.

**Comment by ngaut** [comment]
LGTM

**Pull request merged**